### PR TITLE
return 404 for search when requested format is not html

### PR DESCRIPTION
**Pull request by emrox**
this solves https://github.com/lobsters/lobsters/issues/746 and returns a 404 when the search is requested with a different format than html

room for improvement: don't execute the search when the requested format is not html. This would may require another method to keep the code cleaner and IDK if it's worth the effort

### CODE REVIEW

**Comment by pushcx** [comment]
Running the search when we can't provide the output is painfully wasteful - search is our most expensive page on the site.

Why 404 instead of 400?

Is there a better global solution here, to catch all unavailable formats from the outside? I suspect we may have goofed something in Rails config that this bug can occur at all.

**Comment by emrox** [comment]
> Why 404 instead of 400?

400 signals just "something was wrong here" and could be used as general error. 404 is more specific and tells the remote client the absence of the requested resource. It's like when requesting `file.rss` from a static webserver - when it's not found the webserver returns a 404, even when a `file.html` exists.

**Comment by emrox** [comment]
> Is there a better global solution here

yes, but rss is a valid format for Lobsters, so it may should not be disallowed globally. AFAIK rss is allowed & valid for the most routes so it may make sense to just block it for individual routes. But it also could be done in the other way to not accept it globally and just allow it on certain routes

**Comment by pushcx** [comment]
I was trying to be a bit coy when I asked about a global solution, but to be clearer and put my inbox at risk (I get an email for every site 500) - this is not a problem for this controller action, it's a bug for every route. (I promise it's real, please quit testing this bug in prod, folks.) That's why I suspect we have some kind of Rails misconfiguration.

**Comment by emrox** [comment]
> this is not a problem for this controller action, it's a bug for every route. (I promise it's real, please quit testing this bug in prod, folks.) That's why I suspect we have some kind of Rails misconfiguration.

I'll take a look into it. Should not happen in general

**Comment by emrox** [comment]
updated after merging https://github.com/lobsters/lobsters/pull/765

**Comment by pushcx** [comment]
With #765 merged, what's left to do here?

**Comment by emrox** [comment]
This PR just ensures the search is not executed when requested with an unknown format.

F.e. when someone requests the search with RSS the controller performs all actions just to figure out at the end that the rss view is missing.

File: spec/controllers/search_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+require 'rails_helper'
+
+describe SearchController do
+  describe '#index' do
+    context 'when requesting html view' do
+      it 'responses with oki' do
```
Comment (pushcx) [inline spec/controllers/search_controller_spec.rb]:
```suggestion
      it 'responds with 200' do
```

File: spec/controllers/search_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        get :index
+
+        expect(response).to have_http_status(:ok)
+      end
+    end
+
+    context 'when requested with another format' do
+      it 'responses with a 404' do
```
Comment (pushcx) [inline spec/controllers/search_controller_spec.rb]:
```suggestion
      it 'responds with 404' do
```

File: spec/controllers/search_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        get :index, format: :rss
+
```
Comment (pushcx) [inline spec/controllers/search_controller_spec.rb]: Please add a mock to test that a search didn't run, maybe that `Search.new` or `.search_for_user` wasn't called.

**Comment by pushcx** [comment]
Wait, I left some small notes, but I just realized:

> This PR just ensures the search is not executed when requested with an unknown format.

If this is the case, then this is an issue for *every* action when a missing format is requested. This can't be fixing it at the right layer of abstraction, we must need to fix it in routes or `ApplicationController`. We shouldn't edit every individual action like this.

**Comment by pushcx** [comment]
Hey @emrox, checking in on the PR queue. Any thoughts on how we could solve this at a higher layer for all actions?

**Comment by emrox** [comment]
@pushcx in theory this already merged PR should do the most for all other things? https://github.com/lobsters/lobsters/pull/765

When you just want to have RSS for some routes it may should be solved in a different way. But I'm not sure. Maybe just have Html rendered, and RSS only when there is a view file for it?

**Comment by pushcx** [comment]
It still runs the action. If I put `raise "dang"` at the top of `SearchController#index` I get that exception instead of a 404.

**Comment by pushcx** [comment]
It's been a few months - were you still investigating, or should I close this PR and file a new issue to track the broader issue of running the action for unavailable formats?

**Comment by emrox** [comment]
IMHO it can be closed. And it may be an edge case?

**Pull request closed without merging**